File: frontend/src/api/visits.ts
import api from "./client";

export interface KutirVisit {
  id: number;
  kutir_id: number;
  visited_by_id: number | null;
  visit_date: string;
  avg_attendance_last_week: number;
  follow_timetable: boolean;
  follow_monthly_plan: boolean;
  timetable_plan_reason: string | null;
  math_topics_pre: string | null;
  math_topics_upper: string | null;
  english_topics_pre: string | null;
  english_topics_upper: string | null;
  timeslot_utilization: boolean;
  timeslot_reason: string | null;
  grs_prep_remarks: string | null;
  physical_vs_registered: "Matched" | "Not Matched";
  workbook_percentage: number;
  workbook_completion: "Upto Date" | "Partial Upto Date" | "Not Uptodate";
  book_availability: "Sufficient" | "Lacking" | "More than required";
  cleanliness: number;
  hindi_proficiency: number;
  english_proficiency: number;
  maths_proficiency: number;
  evs_proficiency: number;
  reasoning_proficiency: number;
  material_management: number;
  kutir_performance: number;
  reg_admission_forms: boolean;
  reg_attendance_students: boolean;
  reg_daily_activity: boolean;
  reg_observation: boolean;
  reg_students_data: boolean;
  reg_attendance_teachers: boolean;
  reg_students_documents: boolean;
  regular_students: number | null;
  timeslot_bal_sabha: boolean;
  timeslot_sports: boolean;
  timeslot_yoga: boolean;
  timeslot_value_ed: boolean;
  timeslot_gk_map: boolean;
  visit_photo: string | null;
  final_remarks: string | null;
  created_at: string;
  updated_at: string;
}

export type KutirVisitCreate = Omit<KutirVisit, "id" | "visited_by_id" | "visit_photo" | "created_at" | "updated_at">;

export const BLANK_VISIT: KutirVisitCreate = {
  kutir_id: 0,
  visit_date: new Date().toISOString().slice(0, 10),
  avg_attendance_last_week: 0,
  follow_timetable: false,
  follow_monthly_plan: false,
  timetable_plan_reason: null,
  math_topics_pre: null,
  math_topics_upper: null,
  english_topics_pre: null,
  english_topics_upper: null,
  timeslot_utilization: false,
  timeslot_reason: null,
  grs_prep_remarks: null,
  physical_vs_registered: "Matched",
  workbook_percentage: 0,
  workbook_completion: "Upto Date",
  book_availability: "Sufficient",
  cleanliness: 3,
  hindi_proficiency: 3,
  english_proficiency: 3,
  maths_proficiency: 3,
  evs_proficiency: 3,
  reasoning_proficiency: 3,
  material_management: 3,
  kutir_performance: 3,
  reg_admission_forms: false,
  reg_attendance_students: false,
  reg_daily_activity: false,
  reg_observation: false,
  reg_students_data: false,
  reg_attendance_teachers: false,
  reg_students_documents: false,
  regular_students: null,
  timeslot_bal_sabha: false,
  timeslot_sports: false,
  timeslot_yoga: false,
  timeslot_value_ed: false,
  timeslot_gk_map: false,
  final_remarks: null,
};

export async function listVisits(params: { kutir_id?: number; limit?: number } = {}): Promise<KutirVisit[]> {
  const { data } = await api.get("/kutir-visits", { params: { limit: 200, ...params } });
  return data;
}

export async function getVisit(id: number): Promise<KutirVisit> {
  const { data } = await api.get(`/kutir-visits/${id}`);
  return data;
}

export async function createVisit(payload: KutirVisitCreate): Promise<KutirVisit> {
  const { data } = await api.post("/kutir-visits", payload);
  return data;
}

export async function updateVisit(id: number, payload: Partial<KutirVisitCreate>): Promise<KutirVisit> {
  const { data } = await api.patch(`/kutir-visits/${id}`, payload);
  return data;
}

export async function deleteVisit(id: number): Promise<void> {
  await api.delete(`/kutir-visits/${id}`);
}

export async function uploadVisitPhoto(id: number, file: File): Promise<KutirVisit> {
  const form = new FormData();
  form.append("file", file);
  const { data } = await api.post(`/kutir-visits/${id}/photo`, form, {
    headers: { "Content-Type": "multipart/form-data" },
  });
  return data;
}

File: backend/app/schemas/visits.py
from datetime import date, datetime
from typing import Optional
from pydantic import BaseModel, Field


class KutirVisitBase(BaseModel):
    kutir_id: int
    visited_by_id: Optional[int] = None
    visit_date: date

    # Attendance & implementation
    avg_attendance_last_week: int = Field(..., ge=0)
    follow_timetable: bool = False
    follow_monthly_plan: bool = False
    timetable_plan_reason: Optional[str] = None

    # Topics
    math_topics_pre: Optional[str] = None
    math_topics_upper: Optional[str] = None
    english_topics_pre: Optional[str] = None
    english_topics_upper: Optional[str] = None

    # Time slots
    timeslot_utilization: bool = False
    timeslot_reason: Optional[str] = None

    # GRS prep
    grs_prep_remarks: Optional[str] = None

    # Registers & materials
    physical_vs_registered: str          # Matched / Not Matched
    workbook_percentage: int = Field(..., ge=0, le=100)
    workbook_completion: str             # Upto Date / Partial Upto Date / Not Uptodate
    book_availability: str               # Sufficient / Lacking / More than required

    # Ratings 1–5
    cleanliness: int = Field(..., ge=1, le=5)
    hindi_proficiency: int = Field(..., ge=1, le=5)
    english_proficiency: int = Field(..., ge=1, le=5)
    maths_proficiency: int = Field(..., ge=1, le=5)
    evs_proficiency: int = Field(..., ge=1, le=5)
    reasoning_proficiency: int = Field(..., ge=1, le=5)
    material_management: int = Field(..., ge=1, le=5)
    kutir_performance: int = Field(..., ge=1, le=5)

    # Document registers
    reg_admission_forms: bool = False
    reg_attendance_students: bool = False
    reg_daily_activity: bool = False
    reg_observation: bool = False
    reg_students_data: bool = False
    reg_attendance_teachers: bool = False
    reg_students_documents: bool = False

    # Regular students
    regular_students: Optional[int] = None

    # Time slot activities
    timeslot_bal_sabha: bool = False
    timeslot_sports: bool = False
    timeslot_yoga: bool = False
    timeslot_value_ed: bool = False
    timeslot_gk_map: bool = False

    final_remarks: Optional[str] = None


class KutirVisitCreate(KutirVisitBase):
    pass


class KutirVisitUpdate(BaseModel):
    visited_by_id: Optional[int] = None
    visit_date: Optional[date] = None
    avg_attendance_last_week: Optional[int] = Field(None, ge=0)
    follow_timetable: Optional[bool] = None
    follow_monthly_plan: Optional[bool] = None
    timetable_plan_reason: Optional[str] = None
    math_topics_pre: Optional[str] = None
    math_topics_upper: Optional[str] = None
    english_topics_pre: Optional[str] = None
    english_topics_upper: Optional[str] = None
    timeslot_utilization: Optional[bool] = None
    timeslot_reason: Optional[str] = None
    grs_prep_remarks: Optional[str] = None
    physical_vs_registered: Optional[str] = None
    workbook_percentage: Optional[int] = Field(None, ge=0, le=100)
    workbook_completion: Optional[str] = None
    book_availability: Optional[str] = None
    cleanliness: Optional[int] = Field(None, ge=1, le=5)
    hindi_proficiency: Optional[int] = Field(None, ge=1, le=5)
    english_proficiency: Optional[int] = Field(None, ge=1, le=5)
    maths_proficiency: Optional[int] = Field(None, ge=1, le=5)
    evs_proficiency: Optional[int] = Field(None, ge=1, le=5)
    reasoning_proficiency: Optional[int] = Field(None, ge=1, le=5)
    material_management: Optional[int] = Field(None, ge=1, le=5)
    kutir_performance: Optional[int] = Field(None, ge=1, le=5)
    reg_admission_forms: Optional[bool] = None
    reg_attendance_students: Optional[bool] = None
    reg_daily_activity: Optional[bool] = None
    reg_observation: Optional[bool] = None
    reg_students_data: Optional[bool] = None
    reg_attendance_teachers: Optional[bool] = None
    reg_students_documents: Optional[bool] = None
    regular_students: Optional[int] = None
    timeslot_bal_sabha: Optional[bool] = None
    timeslot_sports: Optional[bool] = None
    timeslot_yoga: Optional[bool] = None
    timeslot_value_ed: Optional[bool] = None
    timeslot_gk_map: Optional[bool] = None
    final_remarks: Optional[str] = None


class KutirVisitOut(KutirVisitBase):
    id: int
    visit_photo: Optional[str] = None
    created_at: datetime
    updated_at: datetime
    model_config = {"from_attributes": True}
